### PR TITLE
Fix #2672 - Connect signed PuSH subscription requests to instance domain

### LINKED ISSUE #2672: v1.3+ does not send toots to instances using a separate WEB_DOMAIN
EDIT: #2111 could be causing this. See comment https://github.com/tootsuite/mastodon/issues/2672#issuecomment-298348460.

Original issue:

- - -

Toots from a number of remote instances stopped appearing on my instance. Notifications remain working. Toots originating from my instance reach other instances just fine.

My instance is on v1.3.2, running on a separate WEB_DOMAIN. Logs show nothing particularly suspicious.

The problem started a couple of days ago, around Apr. 29. The problem appears to be instance-wide, not limited to individual users / toots, unlike #2613 . Manually refreshing PuSH, and re-following users did not fix the problem. Instances from which toots are missing (that I know of) range from v1.2.2 to v1.3.2.

Could someone please shed some light on why and how this is happening? Thank you!

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

### PR DESCRIPTION
**Pull request by Gargron**
Resolves #2739

### CODE REVIEW

File: app/workers/pubsubhubbub/distribution_worker.rb [hunk 1]
```diff
@@ -35,16 +35,16 @@ def distribute_hidden!(stream_entries)
...
   def active_subscriptions
-    Subscription.where(account: @account).active.select('id, callback_url')
+    Subscription.where(account: @account).active.select('id, callback_url, domain')
   end
 
-  def allowed_to_receive?(callback_url)
-    @domains.include?(Addressable::URI.parse(callback_url).host)
+  def allowed_to_receive?(callback_url, domain)
+    (!domain.nil? && @domains.include?(domain)) || @domains.include?(Addressable::URI.parse(callback_url).host)
   end
```
Comment (ThibG) [inline app/workers/pubsubhubbub/distribution_worker.rb]: This is not new to this PR, but I think you actually want `.normalized_host` instead of `.host`.

**Comment by Gargron** [comment]
Bonus round: at the cost of backward compatibility, we could require that the domain field be filled, or not send private toots to it - that way we'd get rid of the non-Mastodon instances getting private toots issue. But that's up for debate since it'd exclude outdated Mastodon instances too

**Review by ClearlyClaire — APPROVED** [review]
Apart from one corner case that is not new to this PR, looks good to me.

File: app/workers/pubsubhubbub/distribution_worker.rb (hunk 1 above)
Comment (Gargron) [inline app/workers/pubsubhubbub/distribution_worker.rb]: If I remember correctly, callback_url is normalized before save.

**Comment by Gargron** [comment]
Ah, I didn't adjust the tests.

Also, the method signature having so many params does look bad, but I dunno how to improve it.

File: app/workers/pubsubhubbub/distribution_worker.rb (hunk 1 above)
Comment (ThibG) [inline app/workers/pubsubhubbub/distribution_worker.rb]: Ah, seems to be the case, indeed.

**Comment by ThibG** [comment]
@Gargron I'm not completely against enforcing the subscription requests to be signed for private toot delivery, but I think this should be done in a later release, exactly for the reasons you mentioned.

**Pull request merged**